storagenode/contact/chore.go

```diff
@@ -48,53 +58,108 @@ func (chore *Chore) Run(ctx context.Context) (err error) {
+			err := chore.pingSatellite(ctx, satellite)
+			if err == nil {
+				return nil
+			}
 
-	return chore.Loop.Run(ctx, func(ctx context.Context) error {
-		if err := chore.pingSatellites(ctx); err != nil {
-			chore.log.Error("pingSatellites failed", zap.Error(err))
-		}
-		return nil
-	})
+			interval := backOff
+			retries := 0
+			for {
+				if !sync2.Sleep(ctx, interval) {
+					chore.log.Error("ping satellite failed after " + strconv.Itoa(retries) + " retries timed out")
+					return ctx.Err()
+				}
+				err := chore.pingSatellite(ctx, satellite)
+				retries++
+				if err == nil {
+					return nil
+				}
+				chore.log.Error("ping satellite failed " + strconv.Itoa(retries) + " times")
+				interval *= 2
+				if interval > chore.interval {
+					interval = chore.interval
+				}
+			}
+		})
+	}
+	chore.mu.Unlock()
+	chore.started.Release()
+	return group.Wait()
 }
 
-func (chore *Chore) pingSatellites(ctx context.Context) (err error) {
+func (chore *Chore) pingSatellite(ctx context.Context, id storj.NodeID) (err error) {
 	defer mon.Task()(&ctx)(&err)
-	var group errgroup.Group
 	self := chore.service.Local()
-	satellites := chore.trust.GetSatellites(ctx)
-	for _, satellite := range satellites {
-		satellite := satellite
-		addr, err := chore.trust.GetAddress(ctx, satellite)
-		if err != nil {
-			chore.log.Error("getting satellite address", zap.Error(err))
-			continue
-		}
-		group.Go(func() error {
-			conn, err := chore.dialer.DialAddressID(ctx, addr, satellite)
-			if err != nil {
-				return err
-			}
-			defer func() { err = errs.Combine(err, conn.Close()) }()
+	address, err := chore.trust.GetAddress(ctx, id)
+	if err != nil {
+		return errPingSatellite.New("failed to get satellite address %s: %v", id.String(), err)
+	}
+	conn, err := chore.dialer.DialAddressID(ctx, address, id)
+	if err != nil {
+		return errPingSatellite.New("failed to dial satellite address %s %s: %v", id.String(), address, err)
+	}
+	_, err = conn.NodeClient().CheckIn(ctx, &pb.CheckInRequest{
+		Address:  self.Address.GetAddress(),
+		Version:  &self.Version,
+		Capacity: &self.Capacity,
+		Operator: &self.Operator,
+	})
+	if err != nil {
+		return errPingSatellite.New("failed to check into satellite %s: %v", id.String(), err)
+	}
+	return nil
+}
 
-			_, err = conn.NodeClient().CheckIn(ctx, &pb.CheckInRequest{
-				Address:  self.Address.GetAddress(),
-				Version:  &self.Version,
-				Capacity: &self.Capacity,
-				Operator: &self.Operator,
-			})
+// Pause stops all the cycles in the contact chore
+func (chore *Chore) Pause() {
+	chore.started.Wait(context.TODO())
+	chore.mu.Lock()
+	defer chore.mu.Unlock()
+	for _, cycle := range chore.cycles {
+		cycle.Pause()
+	}
+}
 
-			return err
+// TriggerWait ensures that each cycle is done at least once and waits for completion.
+// If the cycle is currently running it waits for the previous to complete and then runs.
+func (chore *Chore) TriggerWait() {
+	chore.started.Wait(context.TODO())
```

Comment:
Note, the `context.TODO()`-s should still be passed through as arguments. We just used currently to avoid changing more code than necessary.

Reply:
oops! forgot to change that, thanks